src/JustEat.StatsD/TagsFormatters/StatsDTagsFormatter.cs

```diff
@@ -0,0 +1,172 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+using JustEat.StatsD.Buffered;
+
+namespace JustEat.StatsD.TagsFormatters
+{
+    /// <summary>
+    /// A template class to format StatsD tags with configured values.
+    /// </summary>
+    public abstract class StatsDTagsFormatter : IStatsDTagsFormatter
+    {
+        private const int SafeUdpPacketSize = 512;
+
+        [ThreadStatic]
+        private static char[]? _buffer;
+        private static char[] Buffer => _buffer ??= new char[SafeUdpPacketSize];
+
+        private readonly char[] _prefix;
+        private readonly char[] _suffix;
+        private readonly char[] _tagsSeparator;
+        private readonly char[] _keyValueSeparator;
+        private readonly int _prefixSize;
+        private readonly int _suffixSize;
+        private readonly int _tagsSeparatorSize;
+        private readonly int _keyValueSeparatorSize;
+        
+        /// <summary>
+        /// Initializes a new instance of the <see cref="StatsDTagsFormatter"/> class.
+        /// </summary>
+        /// <param name="configuration">The configuration.</param>
+        protected StatsDTagsFormatter(StatsDTagsFormatterConfiguration configuration)
+        {
+            if (configuration == null)
+            {
+                throw new ArgumentNullException(nameof(configuration));
+            }
+
+            _prefix = configuration.Prefix?.ToArray() ?? Array.Empty<char>();
+            _suffix = configuration.Suffix?.ToArray() ?? Array.Empty<char>();
+            AreTrailing = configuration.AreTrailing;
+            _tagsSeparator = configuration.TagsSeparator?.ToArray() ?? Array.Empty<char>();
+            _keyValueSeparator = configuration.KeyValueSeparator?.ToArray() ?? Array.Empty<char>();
+            _prefixSize = Encoding.UTF8.GetByteCount(_prefix);
+            _suffixSize = Encoding.UTF8.GetByteCount(_suffix);
+            _tagsSeparatorSize = Encoding.UTF8.GetByteCount(_tagsSeparator);
+            _keyValueSeparatorSize = Encoding.UTF8.GetByteCount(_keyValueSeparator);
+        }
+        
+        /// <inheritdoc />
+        public bool AreTrailing { get; }
```

Comment:
Might as well make this virtual for consistency.

```suggestion
        public virtual bool AreTrailing { get; }
```

Reply:
Done!